sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -958,6 +963,8 @@ object SparkSession extends Logging {
         sparkContext.addSparkListener(new SparkListener {
           override def onApplicationEnd(applicationEnd: SparkListenerApplicationEnd): Unit = {
             defaultSession.set(null)
+            // Should remove listener after this event fires
+            sparkContext.removeSparkListener(this)
```

Comment:
I think this should be called in the `SparkSession.stop()` method?

Reply:
Yeah, that would be ideal - the problem is that we lost the handle to this listener outside of this method. I could create a global `var` in the `SparkSession` to hold a reference to the listener, but that also seems kind of strange. Thoughts?

Reply:
It should be fine to have a global var of SparkListener inside SparkSession.

Reply:
Okay, will add this in!